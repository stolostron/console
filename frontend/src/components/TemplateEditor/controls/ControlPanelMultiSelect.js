'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { Select, SelectOption, SelectVariant, Spinner } from '@patternfly/react-core'
import ControlPanelFormGroup from './ControlPanelFormGroup'
import get from 'lodash/get'

class ControlPanelMultiSelect extends React.Component {
  static propTypes = {
    control: PropTypes.object,
    controlId: PropTypes.string,
    handleChange: PropTypes.func,
    i18n: PropTypes.func,
  }

  constructor(props) {
    super(props)
    this.state = {
      open: false,
    }
  }

  setControlRef = (control, ref) => {
    this.multiSelect = control.ref = ref
  }

  render() {
    const { open } = this.state
    const { controlId, i18n, control, handleChange } = this.props
    const { available = [], availableMap, exception, disabled, isLoading, isFailed } = control
    let { active, placeholder = '' } = control
    if (!active) {
      if (isLoading) {
        active = i18n(get(control, 'fetchAvailable.loadingDesc', 'resource.loading'))
      } else if (isFailed) {
        active = i18n('resource.error')
      } else if (available.length === 0) {
        active = i18n(get(control, 'fetchAvailable.emptyDesc', 'resource.none'))
      } else {
        active = []
      }
    } else if (Array.isArray(active) && active.length > 0) {
      const activeKeys = []
      active.forEach((k) => {
        if (typeof availableMap === 'object' && availableMap[k]) {
          const { name: n } = availableMap[k]
          activeKeys.push(n || k)
        } else {
          activeKeys.push(k)
        }
      })
      placeholder = activeKeys.join(', ')
    }

    const setOpen = (open) => {
      this.setState({ open })
    }

    const onChange = (value) => {
      if (value) {
        if (active.includes(value)) {
          active = active.filter((item) => item !== value)
        } else {
          active = [...active, value]
        }
      } else {
        active = []
      }
      control.active = active
      handleChange()
    }

    this.options = available.map((item, inx) => {
      /* eslint-disable-next-line react/no-array-index-key */
      return <SelectOption key={inx} value={item} />
    })

    const onFilter = (evt) => {
      const textInput = get(evt, 'target.value', '')
      if (textInput === '') {
        return this.options
      } else {
        return this.options.filter((item) => {
          return item.props.value.toLowerCase().includes(textInput.toLowerCase())
        })
      }
    }

    const validated = exception ? 'error' : undefined
    return (
      <React.Fragment>
        <div className="creation-view-controls-singleselect" ref={this.setControlRef.bind(this, control)}>
          <ControlPanelFormGroup controlId={controlId} control={control}>
            {isLoading ? (
              <div className="creation-view-controls-singleselect-loading">
                <Spinner size="md" />
                <div>{active}</div>
              </div>
            ) : (
              <Select
                ariaLabelledBy={`${controlId}-label`}
                spellCheck={false}
                isOpen={open}
                onToggle={() => {
                  setOpen(!open)
                }}
                variant={SelectVariant.checkbox}
                onSelect={(_event, value) => {
                  onChange(value)
                }}
                selections={active}
                onClear={() => {
                  onChange(undefined)
                }}
                placeholderText={placeholder}
                isDisabled={disabled}
                onFilter={onFilter}
                hasInlineFilter
                data-testid={`multi-${controlId}`}
              >
                {this.options}
              </Select>
            )}
            {validated === 'error' ? (
              <div
                style={{
                  borderTop: '1.75px solid red',
                  paddingBottom: '6px',
                  maxWidth: '600px',
                }}
              ></div>
            ) : (
              <React.Fragment />
            )}
          </ControlPanelFormGroup>
        </div>
      </React.Fragment>
    )
  }
}

export default ControlPanelMultiSelect
