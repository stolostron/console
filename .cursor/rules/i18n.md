# Internationalization Rules for ACM Console

## Translation Patterns

### Hook Usage
- **REQUIRED**: Use acm-i18next library for translations (ESLint enforced)
- **DO NOT** import directly from 'react-i18next' - ESLint will error
- Import and use the translation hook consistently:
```typescript
import { useTranslation } from '../lib/acm-i18next'

export function MyComponent() {
  const { t } = useTranslation()
  return <div>{t('My translated text')}</div>
}
```

### Complex Translations
- **REQUIRED**: Use Trans component from acm-i18next (ESLint enforced)
- **DO NOT** import Trans from 'react-i18next' directly
```typescript
import { Trans } from '../lib/acm-i18next'

<Trans 
  i18nKey="<bold>Warning:</bold> This action cannot be undone."
  components={{ bold: <strong /> }}
/>
```

### Translation Keys
- Use descriptive, hierarchical keys
- Avoid generic keys like 'button', 'text', 'label'
- Use consistent naming patterns
- Group related translations logically

Good examples:
```typescript
t('cluster.create.title')
t('cluster.create.description')
t('cluster.status.ready')
t('cluster.status.notReady')
```

### String Externalization
- All user-facing strings must be externalized
- No hardcoded English text in components
- Use translation keys even for single words
- Include context in translation keys when needed

### Pluralization
- Use i18next pluralization features:
```typescript
t('cluster.count', { count: clusterCount })
// In translation file:
// "cluster.count_one": "{{count}} cluster"
// "cluster.count_other": "{{count}} clusters"
```

### Variable Interpolation
- Use proper variable interpolation:
```typescript
t('cluster.created.success', { clusterName: cluster.name })
// In translation file:
// "cluster.created.success": "Cluster {{clusterName}} created successfully"
```

### Translation Files
- Store all translations in `public/locales/en/translation.json`
- Use nested objects for organization
- Keep translation files well-structured
- Use consistent formatting and indentation

### Testing Translations
- Test that all translation keys exist
- Verify text rendering with different content lengths
- Test variable interpolation works correctly
- Use i18next-parser to extract keys

### Performance
- Avoid creating translation functions in render loops
- Use useMemo for complex translation computations
- Load translations efficiently
- Cache translation results when appropriate

### Accessibility
- Ensure translated text works with screen readers
- Consider text direction for RTL languages
- Test text overflow with longer translations
- Maintain proper semantic structure
