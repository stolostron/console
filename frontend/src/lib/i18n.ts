/* Copyright Contributors to the Open Cluster Management project */

/* istanbul ignore file */

import i18n from 'i18next'
// eslint-disable-next-line @typescript-eslint/no-restricted-imports
import { initReactI18next } from 'react-i18next'
import LanguageDetector from 'i18next-browser-languagedetector'
import HttpApi from 'i18next-http-backend'
import { supportedLanguages } from './supportedLanguages'

i18n
    // pass the i18n instance to react-i18next
    .use(initReactI18next)
    // detect user language
    // learn more: https://github.com/i18next/i18next-browser-languageDetector
    .use(LanguageDetector)
    // fetch json files
    // learn more: https://github.com/i18next/i18next-http-backend
    .use(HttpApi)
    // init i18next
    .init({
        backend: {
            loadPath: '/multicloud/locales/{{lng}}/{{ns}}.json',
        },
        compatibilityJSON: 'v3',
        fallbackLng: ['en'], // if language is not supported or string is missing, fallback to English
        keySeparator: false, // this repo will use single level json
        interpolation: {
            escapeValue: false, // react handles this already
        },
        defaultNS: 'translation', // the default file for strings when using useTranslation, etc
        nsSeparator: '~',
        supportedLngs: supportedLanguages, // only languages from this array will attempt to be loaded
        simplifyPluralSuffix: true,
    })

export default i18n
