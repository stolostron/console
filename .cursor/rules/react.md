# React Rules for ACM Console

## Component Patterns

### Functional Components
- Use functional components with hooks exclusively
- Implement proper prop interfaces with TypeScript
- Use React.memo for expensive components
- Follow established naming conventions (PascalCase)

### Hook Usage
- Use custom hooks for business logic separation
- Implement proper dependency arrays in useEffect
- Use useCallback for event handlers passed as props
- Leverage useMemo for expensive computations

### State Management
- Use Recoil atoms through useSharedAtoms() pattern
- Access atoms with useRecoilValue() for reading
- Use useRecoilValueGetter() for async operations
- Follow established atom naming conventions

Example (based on actual codebase patterns):
```typescript
/* Copyright Contributors to the Open Cluster Management project */
import { useMemo } from 'react'
import { useRecoilValue } from 'recoil'
import { useTranslation } from '../lib/acm-i18next'
import { useSharedAtoms } from '../shared-recoil'
import { AcmTable } from '../ui-components'

export function ClusterList() {
  const { t } = useTranslation()
  const { managedClustersState } = useSharedAtoms()
  const clusters = useRecoilValue(managedClustersState)
  
  const filteredClusters = useMemo(
    () => clusters.filter(cluster => cluster.status?.conditions),
    [clusters]
  )
  
  return (
    <AcmTable
      items={filteredClusters}
      columns={clusterColumns}
      keyFn={cluster => cluster.metadata.uid}
    />
  )
}
```

### Conditional Rendering
- Use logical AND (`&&`) for simple conditional rendering (ACM standard pattern):
  ```typescript
  {condition && <Component />}
  {props.additionalLabels && (<span><AcmLabels /></span>)}
  {expandable && itemCount > columnCount && (<AcmButton />)}
  ```
- Use ternary operators for either/or scenarios:
  ```typescript
  {props.isCompact ? <Popover /> : renderLabelGroup()}
  {showAll ? t('Show less') : t('Show all')}
  ```
- Early returns for major conditional logic:
  ```typescript
  if (props.hidden) return <></>
  if (!data) return <LoadingState />
  ```

### Performance
- Use `React.memo` for components that receive the same props frequently
- Use `useMemo` for expensive calculations
- Use `useCallback` for functions passed to child components
- Implement lazy loading for heavy components
- Avoid inline object/array creation in render
- Avoid using index as the key prop value
- Avoid using inline functions in JSX or TSX blocks
- Use proper key props for list items

### Error Boundaries
- Implement error boundaries for route-level components
- Use proper error state management
- Provide user-friendly error messages
- Log errors for debugging purposes

### Accessibility
- Use semantic HTML elements
- Implement proper ARIA attributes
- Ensure keyboard navigation works
- Test with screen readers

### Component Organization
- Keep components small and focused on a single responsibility
- Use PascalCase for component names and component files
- Use descriptive component names that indicate the component's purpose
- Avoid abbreviations in component names
- Separate presentation from business logic
- Use proper file structure (components/, ui-components/, routes/)
- Export components properly with named exports
- Avoid props drilling where props are only passed down to child components
- Do not mutate props or state - create a copy array or object if possible
