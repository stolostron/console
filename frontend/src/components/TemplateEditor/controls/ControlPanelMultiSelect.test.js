/* Copyright Contributors to the Open Cluster Management project */
'use strict'

import React from 'react'
import ControlPanelMultiSelect from './ControlPanelMultiSelect'
import { render } from '@testing-library/react'

import i18n from 'i18next'

const t = i18n.t.bind(i18n)

export const control = {
  active: false,
  name: 'creation.app.name',
  tooltip: 'tooltip.creation.app.name',
  controlData: [],
  id: 'checkbox',
  type: 'checkbox',
}

describe('ControlPanelMultiSelect component', () => {
  it('renders as expected', () => {
    const fn = jest.fn()

    const Component = () => {
      return (
        <ControlPanelMultiSelect key={'key'} control={control} controlId={'controlId'} handleChange={fn} i18n={t} />
      )
    }
    const { asFragment } = render(<Component />)
    expect(asFragment()).toMatchSnapshot()
  })
})
