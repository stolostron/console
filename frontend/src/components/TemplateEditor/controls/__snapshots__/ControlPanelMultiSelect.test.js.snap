// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlPanelMultiSelect component renders as expected 1`] = `
<DocumentFragment>
  <div
    class="creation-view-controls-singleselect"
  >
    <div>
      <div
        class="pf-c-form__group"
        id="controlId-label"
      >
        <div
          class="pf-c-form__group-label"
        >
          <label
            class="pf-c-form__label"
            for="controlId"
          >
            <span
              class="pf-c-form__label-text"
            >
              creation.app.name
            </span>
          </label>
           
          <button
            aria-label="More info"
            class="pf-c-form__group-label-help"
            id="controlId-label-help-button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
              />
            </svg>
          </button>
        </div>
        <div
          class="pf-c-form__group-control"
        >
          <div
            class="pf-c-select"
            data-ouia-component-id="OUIA-Generated-Select-checkbox-1"
            data-ouia-component-type="PF4/Select"
            data-ouia-safe="true"
          >
            <div
              class="pf-c-select__toggle pf-m-typeahead"
            >
              <div
                class="pf-c-select__toggle-wrapper"
              >
                <span
                  class="pf-c-select__toggle-text"
                />
                <div
                  class="pf-c-select__toggle-badge"
                >
                  <span
                    class="pf-c-badge pf-m-read"
                  />
                </div>
              </div>
              <button
                aria-label="Clear all"
                class="pf-c-button pf-m-plain pf-c-select__toggle-clear"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
                  />
                </svg>
              </button>
              <button
                aria-expanded="false"
                aria-label="Options menu"
                aria-labelledby=" pf-select-toggle-id-0"
                class="pf-c-button pf-c-select__toggle-button pf-m-plain"
                id="pf-select-toggle-id-0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="pf-c-select__toggle-arrow"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 320 512"
                  width="1em"
                >
                  <path
                    d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
